src/wp-includes/class-wp-theme-json.php

```diff
@@ -2124,6 +2159,48 @@ public function get_styles_for_block( $block_metadata ) {
 			}
 		}
 
+		// If there are style variations, generate the declarations for them, including any feature selectors the block may have.
+		$style_variation_declarations = array();
+		if ( ! empty( $block_metadata['variations'] ) ) {
+			foreach ( $block_metadata['variations'] as $style_variation ) {
+				$style_variation_node     = _wp_array_get( $this->theme_json, $style_variation['path'], array() );
+				$style_variation_selector = $style_variation['selector'];
+
+				// If the block has feature selectors, generate the declarations for them within the current style variation.
+				if ( ! empty( $block_metadata['features'] ) ) {
+					foreach ( $block_metadata['features'] as $feature_name => $feature_selector ) {
+						if ( ! empty( $style_variation_node[ $feature_name ] ) ) {
+							// Prepend the variation selector to the feature selector.
+							$split_feature_selectors    = explode( ',', $feature_selector );
+							$feature_selectors          = array_map(
+								function( $split_feature_selector ) use ( $style_variation_selector ) {
+									return trim( $style_variation_selector ) . trim( $split_feature_selector );
```

Comment:
A few things here for micro-optimizations:

* Use `static function`. That way the callback is created once rather than on each iteration.

* Move the `trim( $style_variation_selector )` outside of the `foreach`. It does not change without the loop. By moving it outside of the loop, trimming happens once instead of on each iteration.

Reply:
Resolved here https://github.com/WordPress/wordpress-develop/pull/3941/commits/994c20294df6ea20c57354b309fbcee48485eb9e.